include/deal.II/lac/block_direct_solver.h

```diff
@@ -0,0 +1,407 @@
+  const unsigned int end_range = owned_dofs[n_blocks - 1].nth_index_in_set(0) +
+                                 dof_shifts[this_mpi_process][n_blocks - 1] +
+                                 owned_dofs[n_blocks - 1].n_elements();
+
+  shifted_owned.add_range(begin_range, end_range);
+
+
+  n_local_dofs = 0;
+  for (unsigned int b = 0; b < n_blocks; ++b)
+    n_local_dofs += owned_dofs[b].n_elements();
+
+  AssertDimension(n_local_dofs, shifted_owned.n_elements());
+
+  Assert(shifted_owned.is_ascending_and_one_to_one(mpi_comm),
+         ExcMessage(
+           "The shifted dofs are not valid index set (internal error)"));
+
+
+  first_block_index.resize(n_blocks);
+  last_block_index.resize(n_blocks);
+  for (unsigned int b = 0; b < n_blocks; ++b)
+    {
+      first_block_index[b] =
+        dof_shifts[this_mpi_process][b] + owned_dofs[b].nth_index_in_set(0);
+      last_block_index[b] =
+        first_block_index[b] + owned_dofs[b].n_elements() - 1;
+    }
+
+  AssertDimension(last_block_index[n_blocks - 1],
+                  first_block_index[0] + n_local_dofs - 1);
+}
+
+
+
+template <typename LevelNumber>
+template <class Operator>
+void
+BlockDirectSolver<LevelNumber>::initialize(const Operator &    vmult_operator,
+                                           const unsigned int &lvl)
+{
+  matrix = std::make_unique<TrilinosWrappers::SparseMatrix>(shifted_owned,
+                                                            shifted_owned,
+                                                            mpi_comm,
+                                                            100);
+  level  = lvl;
+
+  BlockVectorType dst;
+  BlockVectorType src;
+
+  src.reinit(n_blocks);
+  dst.reinit(n_blocks);
+  for (unsigned int b = 0; b < n_blocks; ++b)
+    {
+      src.block(b).reinit(owned_dofs[b], relevant_dofs[b], mpi_comm);
+      dst.block(b).reinit(owned_dofs[b], relevant_dofs[b], mpi_comm);
+    }
+  src.collect_sizes();
+  dst.collect_sizes();
+
+  for (unsigned int b = 0; b < n_blocks; ++b)
+    {
+      for (types::global_dof_index i = 0; i < owned_dofs[b].size(); ++i)
+        {
+          src = 0;
+          dst = 0;
+          if (owned_dofs[b].is_element(i))
+            {
+              src.block(b)(i) = 1;
+            }
+
+          src.compress(VectorOperation::insert);
+          vmult_operator.vmult(dst, src);
+
+          dst.update_ghost_values();
+
+
+          for (unsigned int k = 0; k < n_blocks; ++k)
+            {
+              for (IndexSet::ElementIterator index_iter = owned_dofs[k].begin();
+                   index_iter != owned_dofs[k].end();
+                   ++index_iter)
+                {
+                  if (dst.block(k)(*index_iter) != 0)
+                    matrix->set(ext2int(*index_iter, k),
+                                ext2int(i, b),
+                                dst.block(k)(*index_iter));
+                }
+            }
+        }
+    }
+  matrix->compress(VectorOperation::unknown);
```

Comment:
@mwichro Would it possible to extract this into separate that is designated to compute from an operator a block matrix?

Reply:
The point is I am not computing the BlockMatrix but standard SparseMatrix. The internal DoF renumbering is the opposite what DoFRenumbering::component_wise does. The vmult operation is defined as PA^-1 P^-1, where A is internally stored matrix, and P is permutation matrix that ensures the locally owned index set are contiguous. 

For efficiently computing BlockMatrix from operator and solving linear system I would go other way around: first compute BlockMatrix using a sparsity pattern ( PR #9723 ), and than initialize this solver with BlockMatrix (functionality to be added here). Since for some time you will be storing both original and permutated matrix this will be not memmory efficient, but the main purpose of this solver is solving coarse problem (that usually signifficantly smaller that original and thus we don't have to care about memmory so much) and checking if your matrix/operator is properly assembled (looking at the solution of linear system may help localize the problem). In second case the user would most probably be debugging the code -- if you are using block matrices you are probably going to use iterative solver.